### PR TITLE
Sync ``libasr`` from LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
LFortran PR - https://github.com/lfortran/lfortran/pull/2403

### CODE REVIEW

File: src/libasr/asr_utils.h [hunk 1]
```diff
@@ -1091,15 +1131,15 @@ static inline std::string extract_dim_value(ASR::expr_t* dim) {
...
+        dims_str += "[";
+        dims_str += extract_dim_value(dim.m_start);
+        dims_str += ",";
         dims_str += extract_dim_value(dim.m_length);
-        if (i + 1 < n_dims) {
-            dims_str += ",";
-        }
+        dims_str += "]";
     }
-    dims_str += "]";
```
Comment (Shaikh-Ubaid) [inline src/libasr/asr_utils.h:1102]: I think this needs to be reverted (https://github.com/lcompilers/lpython/commit/fdde3c924f5b0b129392bdd3d8a089d71e169bc0).

File: src/libasr/asr_utils.cpp [hunk 2]
```diff
@@ -1266,13 +1266,13 @@ ASR::asr_t* make_Cast_t_value(Allocator &al, const Location &a_loc,
                                 ASRUtils::expr_value(a_arg))->m_n;
             value = ASR::down_cast<ASR::expr_t>(ASR::make_UnsignedIntegerConstant_t(al, a_loc, int_value, a_type));
         } else if (a_kind == ASR::cast_kindType::UnsignedIntegerToInteger) {
-            int64_t int_value = ASR::down_cast<ASR::UnsignedIntegerConstant_t>(
-                                ASRUtils::expr_value(a_arg))->m_n;
-            value = ASR::down_cast<ASR::expr_t>(ASR::make_IntegerConstant_t(al, a_loc, int_value, a_type));
+             int64_t int_value = ASR::down_cast<ASR::UnsignedIntegerConstant_t>(
+                                 ASRUtils::expr_value(a_arg))->m_n;
+             value = ASR::down_cast<ASR::expr_t>(ASR::make_IntegerConstant_t(al, a_loc, int_value, a_type));
```
Comment (certik) [inline src/libasr/asr_utils.cpp:1271]: I think the identation is incorrect here, it should be 4 spaces.

File: src/libasr/asr_utils.cpp [hunk 3]
```diff
@@ -1266,13 +1266,13 @@ ASR::asr_t* make_Cast_t_value(Allocator &al, const Location &a_loc,
...
         } else if (a_kind == ASR::cast_kindType::UnsignedIntegerToUnsignedInteger) {
-            int64_t int_value = ASR::down_cast<ASR::UnsignedIntegerConstant_t>(
-                                ASRUtils::expr_value(a_arg))->m_n;
-            value = ASR::down_cast<ASR::expr_t>(ASR::make_UnsignedIntegerConstant_t(al, a_loc, int_value, a_type));
+             int64_t int_value = ASR::down_cast<ASR::UnsignedIntegerConstant_t>(
+                                 ASRUtils::expr_value(a_arg))->m_n;
+             value = ASR::down_cast<ASR::expr_t>(ASR::make_UnsignedIntegerConstant_t(al, a_loc, int_value, a_type));
```
Comment (certik) [inline src/libasr/asr_utils.cpp:1275]: incorrect indentation

**Comment by ansharlubis** [comment]
`instantiate_template` looks good to me!

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 4]
```diff
@@ -8318,22 +8902,37 @@ Result<std::unique_ptr<LLVMModule>> asr_to_llvm(ASR::TranslationUnit_t &asr,
...
+    pass_options.module_name_mangling = co.module_name_mangling;
+    pass_options.global_symbols_mangling = co.global_symbols_mangling;
+    pass_options.intrinsic_symbols_mangling = co.intrinsic_symbols_mangling;
+    pass_options.bindc_mangling = co.bindc_mangling;
+    pass_options.mangle_underscore = co.mangle_underscore;
     pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
 
     // Uncomment for debugging the ASR after the transformation
-    // std::cout << LCompilers::LPython::pickle(asr, true, true, false) << std::endl;
+    // std::cout << LFortran::pickle(asr, true, true, false) << std::endl;
```
Comment (Pranavchiku) [inline src/libasr/codegen/asr_to_llvm.cpp:8935]:
```suggestion
    // std::cout << LCompilers::pickle(asr, true, true, false) << std::endl;
```

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 5]
```diff
@@ -469,16 +504,18 @@ R"(#include <stdio.h>
                     if (gen_stdcomplex) {
                         sub = "std::complex<float> ";
                     } else {
-                        sub = "float_complex_t ";
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:472]: This change should be reverted. See: https://github.com/lcompilers/lpython/pull/2193/files

File: integration_tests/CMakeLists.txt [hunk 6]
```diff
@@ -723,7 +723,7 @@ RUN(NAME enum_06             LABELS cpython llvm c)
 RUN(NAME enum_07             IMPORT_PATH ..
                              LABELS cpython llvm c)
 RUN(NAME union_01            LABELS cpython llvm c)
-RUN(NAME union_02            LABELS cpython llvm c)
+RUN(NAME union_02            LABELS cpython llvm c NOFAST)
```
Comment (czgdp1807) [inline integration_tests/CMakeLists.txt:726]: This test silently works incorrectly on ``main`` branch as well.

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 7]
```diff
@@ -836,15 +836,14 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
     void visit_Program(const ASR::Program_t &x) {
         // Generate the bodies of functions and subroutines
         declare_all_functions(*x.m_symtab);
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:838]: From https://github.com/lfortran/lfortran/pull/2403#discussion_r1321991515

```suggestion
```

File: integration_tests/CMakeLists.txt (hunk 6 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:726]: How so? How can it silently work? Is our cmake flaky? If so, that's an urgent bug to fix.

File: src/libasr/codegen/asr_to_c.cpp [hunk 8]
```diff
@@ -1061,10 +1070,19 @@ R"(    // Initialise Numpy
...
+                }
                 if (ASR::is_a<ASR::Complex_t>(*value_type)) {
                     tmp_gen += "creal(" + src + ")";
                     tmp_gen += ", ";
                     tmp_gen += "cimag(" + src + ")";
+                } else if(ASR::is_a<ASR::SymbolicExpression_t>(*value_type)){
+                    tmp_gen += "basic_str(" + src + ")";
+                    if(ASR::is_a<ASR::Var_t>(*x.m_msg)) {
+                        symengine_queue.pop();
+                    }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c.cpp:1085]: I think this is to be reverted (a554bbd327c6ce97e35f4923f716a9c5ea737b50)

File: src/libasr/codegen/asr_to_c.cpp [hunk 9]
```diff
@@ -1139,6 +1157,10 @@ R"(    // Initialise Numpy
             if( ASRUtils::is_array(value_type) ) {
                 src += "->data";
             }
+            if(ASR::is_a<ASR::SymbolicExpression_t>(*value_type)) {
+                out += symengine_src;
+                symengine_src = "";
+            }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c.cpp:1163]: To be reverted (a554bbd327c6ce97e35f4923f716a9c5ea737b50)

File: src/libasr/codegen/asr_to_c.cpp [hunk 10]
```diff
@@ -1161,6 +1183,12 @@ R"(    // Initialise Numpy
                 v.pop_back();
                 v.push_back("creal(" + src + ")");
                 v.push_back("cimag(" + src + ")");
+            } else if(ASR::is_a<ASR::SymbolicExpression_t>(*value_type)){
+                v.pop_back();
+                v.push_back("basic_str(" + src + ")");
+                if(ASR::is_a<ASR::Var_t>(*x.m_values[i])) {
+                    symengine_queue.pop();
+                }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c.cpp:1191]: TBR (a554bbd327c6ce97e35f4923f716a9c5ea737b50)

**Comment by Shaikh-Ubaid** [comment]
I think all the changes related to https://github.com/lcompilers/lpython/pull/2260/files (there are several in this PR currently) are to be reverted.

**Comment by Shaikh-Ubaid** [comment]
>I think all the changes related to https://github.com/lcompilers/lpython/pull/2260/files (there are several in this PR currently) are to be reverted.

They might need to be updated to LFortran PR as well.

**Review by Thirumalai-Shaktivel — APPROVED** [review]
LGTM! Thank you very much!

File: integration_tests/CMakeLists.txt (hunk 6 above)
Comment (czgdp1807) [reply integration_tests/CMakeLists.txt:726]: With `main` branch,

```zsh
(lp) 14:17:37:~/lpython_project/lpython % lpython integration_tests/union_02.py --fast
0 1.00000000000000000e+00
2 7.00000000000000000e+00
5 1.30000000000000000e+01 3.04899892523928701e-314
(lp) 14:17:55:~/lpython_project/lpython % lpython integration_tests/union_02.py       
0 1.00000000000000000e+00
2 7.00000000000000000e+00
5 1.30000000000000000e+01 8.00000000000000000e+00
```

You can clearly see the difference in the last value of third line. `--fast` should have given assertion error in `main` branch but it didn't. So, clearly its passing silently in `main` branch as well. In my `libasr_sync_06` this failure just shows up as an assertion error, something which should have happened in `main` branch only. I will fix it later, after getting this PR merged.

**Pull request merged**